Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift

```diff
@@ -0,0 +1,205 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import _SwiftSyntaxTestSupport
+import XCTest
+import SwiftDiagnostics
+import SwiftParser
+import SwiftParserDiagnostics
+import SwiftSyntax
+
+struct SimpleDiagnosticMessage: DiagnosticMessage {
+  let message: String
+  let diagnosticID: MessageID
+  let severity: DiagnosticSeverity
+}
+
+extension SimpleDiagnosticMessage: FixItMessage {
+  var fixItID: MessageID { diagnosticID }
+}
+
+extension GroupedDiagnostics {
+  /// Add a new test file to the group, starting with marked source and using
+  /// the markers to add any suggested extra diagnostics at the marker
+  /// positions.
+  mutating func addTestFile(
+    _ markedSource: String,
+    displayName: String,
+    parent: (SourceFileID, AbsolutePosition)? = nil,
+    extraDiagnostics: [String: (String, DiagnosticSeverity)] = [:]
+  ) -> (SourceFileID, [String: AbsolutePosition]) {
+    // Parse the source file and produce parser diagnostics.
+    let (markers, source) = extractMarkers(markedSource)
+    let tree = Parser.parse(source: source)
+    var diagnostics = ParseDiagnosticsGenerator.diagnostics(for: tree)
+
+    // Add on any extra diagnostics provided, at their marker locations.
+    for (marker, (message, severity)) in extraDiagnostics {
+      let pos = AbsolutePosition(utf8Offset: markers[marker]!)
+      let node = tree.token(at: pos)!.parent!
+
+      let diag = Diagnostic(
+        node: node,
+        message: SimpleDiagnosticMessage(
+          message: message,
+          diagnosticID: MessageID(domain: "test", id: "conjured"),
+          severity: severity
+        )
+      )
+      diagnostics.append(diag)
+    }
+
+    let id = addSourceFile(
+      tree: tree,
+      displayName: displayName,
+      parent: parent,
+      diagnostics: diagnostics
+    )
+
+    let markersWithAbsPositions = markers.map { (marker, pos) in
+      (marker, AbsolutePosition(utf8Offset: pos))
+    }
+
+    return (id, Dictionary(uniqueKeysWithValues: markersWithAbsPositions))
+  }
+}
+
+final class GroupedDiagnosticsFormatterTests: XCTestCase {
+  func testGroupingForMacroExpansion() {
+    var group = GroupedDiagnostics()
+
+    // Main source file.
+    let (mainSourceID, mainSourceMarkers) = group.addTestFile(
+      """
+      let pi = 3.14159
+      0️⃣#myAssert(pi == 3)
```

Comment:
Not really important but in all the other tests we start counting diagnostics with 1️⃣ not 0️⃣.